lib/palettes_base.cpp

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-443271.
+//
+// This file is part of the GLVis visualization tool and library. For more
+// information and source code availability see https://glvis.org.
+//
+// GLVis is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "palettes_base.hpp"
+#include "palettes_default.cpp"
+
+
+void RGBAf::Print(ostream& os) const
+{
+   os << fixed << setprecision(6)
+      << setw(10) << r << " "
+      << setw(10) << g << " "
+      << setw(10) << b << " "
+      << setw(10) << a;
+}
+
+template <size_t N>
+Palette::Palette(const string& name,
+                 const array<array<float,3>,N>& arr) : name(name)
+{
+   colors.resize(N);
+   for (size_t i = 0; i < N; ++i)
+   {
+      colors[i] = RGBAf(arr[i][0], arr[i][1], arr[i][2]);
+   }
+}
+
+template <size_t N>
+Palette::Palette(const string& name,
+                 const array<array<float,4>,N>& arr) : name(name)
+{
+   colors.resize(N);
+   for (size_t i = 0; i < N; ++i)
+   {
+      colors[i] = RGBAf(arr[i][0], arr[i][1], arr[i][2], arr[i][3]);
+   }
+}
+
+
+void Palette::AddColor(float r, float g, float b, float a)
+{
+   colors.push_back(RGBAf(r, g, b, a));
+}
+
+
+void Palette::Print(ostream& os) const
+{
+   os << "palette " << name << " RGBf" << endl;
+   for (const auto& color : colors)
+   {
+      color.Print(os);
+      os << endl;
+   }
+   os << endl;
+}
+
+RGBAf Palette::Color(int i, bool reversed) const
+{
+   int j = reversed ? Size() - 1 - i : i;
+   return colors[j];
+}
+
+vector<array<float,4>> Palette::GetData(bool reversed) const
+{
+   vector<array<float,4>> rgba_data(Size());
+   for (int i = 0; i < Size(); ++i)
+   {
+      rgba_data[i] = Color(i, reversed).AsArray();
+   }
+   return rgba_data;
+}
+
+bool Palette::IsTranslucent() const
+{
+   for (const auto& color : colors)
+   {
+      if (color.a != 1.0) { return true; }
+   }
+   return false;
+}
+
+// Initialize
+int Texture::max_texture_size = 4096;
```

Comment:
The init below will not work with a non-zero value 🙃 . Besides, I would prefer something really invalid like `-1` to make it obvious.

Reply:
oops; I meant to set that to 0 but was testing and forgot to revert.  I'll change to -1

Reply:
Better, but change the init condition accordingly 🙂